deepinv/datasets/kohler.py

```diff
@@ -0,0 +1,236 @@
+            sharp_frame, blurry_shot = dataset.get_item(1, 1, frame="middle")
+    """
+
+    # The Köhler dataset is split into multiple archives available online.
+    archive_urls = [
+        "http://people.kyb.tuebingen.mpg.de/rolfk/BenchmarkECCV2012/GroundTruth_pngs_Image1.zip",
+        "http://people.kyb.tuebingen.mpg.de/rolfk/BenchmarkECCV2012/GroundTruth_pngs_Image2.zip",
+        "http://people.kyb.tuebingen.mpg.de/rolfk/BenchmarkECCV2012/GroundTruth_pngs_Image3.zip",
+        "http://people.kyb.tuebingen.mpg.de/rolfk/BenchmarkECCV2012/GroundTruth_pngs_Image4.zip",
+        "http://people.kyb.tuebingen.mpg.de/rolfk/BenchmarkECCV2012/BlurryImages.zip",
+    ]
+
+    # The checksums are used to verify the integrity of the downloaded
+    # archives.
+    archive_checksums = {
+        "GroundTruth_pngs_Image1.zip": "acb90b6d9bfdb4b2370e08a5fcb80e68",
+        "GroundTruth_pngs_Image2.zip": "da440d3bf43b32bec0b7170ccd828f29",
+        "GroundTruth_pngs_Image3.zip": "3a77c41c951367f35db52eb18496bbac",
+        "GroundTruth_pngs_Image4.zip": "72ce9690c3ed1296358653396cf9576d",
+        "BlurryImages.zip": "61ffb1434d93fca6c508976a7216d723",
+    }
+
+    # Most of the acquisitions of sharp images span exactly 199 frames but not
+    # all of them and this lookup table gives each frame count for them all.
+    frame_count_table = {
+        (2, 11): 200,
+        (1, 10): 198,
+        (1, 12): 198,
+        (2, 10): 198,
+        (3, 7): 198,
+        (3, 12): 198,
+        (4, 12): 198,
+        "others": 199,
+    }
+
+    def __init__(
+        self,
+        root: Union[str, Path],
+        frame_specifier: Union[int, str] = "middle",
+        ordering: str = "printout_first",
+        transform: Callable = None,
+        download: bool = False,
+    ) -> None:
+        self.root = root
+        self.frame_specifier = frame_specifier
+        self.ordering = ordering
+        self.transform = transform
+
+        if download:
+            self.download(self.root)
+
+    @classmethod
+    def download(cls, root: Union[str, Path], remove_finished: bool = False) -> None:
+        """Download the dataset.
+
+        :param Union[str, Path] root: Root directory of the dataset.
+        :param bool remove_finished: Remove the archives after extraction.
+
+        |sep|
+
+        :Examples:
+
+            Download the dataset ::
+
+                from deepinv.datasets import Kohler
+                Kohler.download("datasets/Kohler")
+        """
+        for url in cls.archive_urls:
+            archive_name = url_basename(url)
+            checksum = cls.archive_checksums[archive_name]
+
+            # Download the archive and verify its integrity
+            download_and_extract_archive(
+                url,
+                root,
+                filename=archive_name,
+                md5=checksum,
+                remove_finished=remove_finished,
+            )
+
+    def __len__(self) -> int:
+        return 48
```

Comment:
In the current implementation, the method `__len__` is consistent with the method `__getitem__`  which indeed returns one of 48 items. The 48 items are a subset of about 9600 possible pairings which depend on the attribute `frame_specifier`. The idea behind this is to have a sensible default interface providing non-redundant pairs of blurry and sharp images. It makes it possible to use all of the tooling provided by deepinv (e.g. `deepinv.test`) without making the same computations hundreds of times. If for some reason someone wants to compare their deblurred images to all of the sharp frames, they can use the nonstandard interface provided by the method `get_item`. Do you think it'd be better to provide all possible pairs in the default interface even though they're very redundant? Another option is to return entire lists instead of pairs, each containing a blurry image and its 200 associated sharp frames.

Reply:
If it makes more sense to work on only a single sharp img per blurry img, I think it is better to keep it like this.

But maybe give the possibility to specify a list of frames which will enable to have `several sharp imgs per blurry img`. I don't have a specific use-case in mind, so it may not be relevant.

Reply:
That's a good point. What do you think of the implementation in 7dea72f?

Reply:
Do you think you could dynamically set len so the test has something to test e.g.
```python
def __len__(self):
    return len(glob(join(self.root, f"Blurry*.png")))
```

Reply:
I'd rather hardcode the number of entries and add an integrity check somewhere than use the file structure as a source of truth. How does it sound?

PS we can't easily test this in CI as there's not enough space for downloading the (whole) dataset

Reply:
Makes sense, thanks!